acceptance_modelisation/base_acceptance_map_creator.py

```diff
@@ -436,19 +451,47 @@ def _create_model_cos_zenith_binned(self,
+        if self.livetime_condition:
+            cut_variable_weights = livetime_observations
+        elif "min_n_observation" in self.cos_zenith_binning_method:
+            cut_variable_weights = np.ones(len(cos_zenith_observations), dtype=int)
+        else:
+            logging.error(f"No {self.cos_zenith_binning_method} method available for the cos zenith binning")
 
-        # Adapt binning to have a given number of run in each bin
+        min_cut_per_cos_zenith_bin = self.cos_zenith_binning_parameter_value
+        cut_variable_per_bin = np.histogram(cos_zenith_observations, bins=cos_zenith_bin, weights=cut_variable_weights)[
+            0]
+        # Adapt binning to have a given number of minimum livetime or run in each bin
         i = 0
-        while i < len(run_per_bin):
-            if run_per_bin[i] < self.min_observation_per_cos_zenith_bin and (i + 1) < len(run_per_bin):
-                run_per_bin[i] += run_per_bin[i + 1]
-                run_per_bin = np.delete(run_per_bin, i + 1)
-                cos_zenith_bin = np.delete(cos_zenith_bin, i + 1)
-            elif run_per_bin[i] < self.min_observation_per_cos_zenith_bin and (i + 1) == len(run_per_bin) and i > 0:
-                run_per_bin[i - 1] += run_per_bin[i]
-                run_per_bin = np.delete(run_per_bin, i)
+        
+        while i < len(cut_variable_per_bin):
+            in_coszd_bin = (cos_zenith_observations >= cos_zenith_bin[i]) & (
+                    cos_zenith_observations < cos_zenith_bin[i + 1])
+            if self.wobble_condition: 
+                wobble_in_bin = wobble_observations[in_coszd_bin]
+                cut_variable_in_bin = cut_variable_weights[in_coszd_bin]
+                wobble_livetime_in_bin = np.array(
+                    [cut_variable_in_bin[wobble_in_bin == wobble].sum() for wobble in np.unique(wobble_in_bin)])
+                at_least_2_wobble = len(np.unique(wobble_in_bin)) > 1
+                if at_least_2_wobble:
+                    livetime_min_for_each_wobble = np.all(
+                        wobble_livetime_in_bin >= min_cut_per_cos_zenith_bin )
+            
+            if ((self.wobble_condition and ((not at_least_2_wobble) or (not livetime_min_for_each_wobble)))
+                or (cut_variable_per_bin[i] < min_cut_per_cos_zenith_bin)) and (i + 1) < len(cut_variable_per_bin):
+                    cut_variable_per_bin[i] += cut_variable_per_bin[i + 1]
+                    cut_variable_per_bin = np.delete(cut_variable_per_bin, i + 1)
+                    cos_zenith_bin = np.delete(cos_zenith_bin, i + 1)
+                
+            elif cut_variable_per_bin[i] < min_cut_per_cos_zenith_bin and (i + 1) == len(
+                    cut_variable_per_bin) and i > 0:
+                cut_variable_per_bin[i - 1] += cut_variable_per_bin[i]
+                cut_variable_per_bin = np.delete(cut_variable_per_bin, i)
```

Comment:
There may be a cleaner way to do all this. I will have a look.

Reply:
I tried to get the boolean names and logics clearer, let me know what you think.

I also realized that the last elif (line 491) needed to include the wobble condition. It is easier to see in the new version.